File: ios/Classes/SignapPlugin.swift
import Flutter
import Foundation
import Signap

/// Flutter plugin for the Signap iOS SDK.
///
/// Marshals the Dart `identify(config, options)` call to the native
/// `Signap` SDK (sdks/ios) — signal collection, derived ids, cert
/// pinning and the `/v1/identify` transport all run there. This plugin only
/// translates the channel dictionary ↔ the native types and bridges async ↔ the
/// Flutter result callback.
public class SignapPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "signap_signals", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(SignapPlugin(), channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard call.method == "identify" else {
            result(FlutterMethodNotImplemented)
            return
        }
        guard
            let args = call.arguments as? [String: Any],
            let config = args["config"] as? [String: Any],
            let apiKey = config["apiKey"] as? String
        else {
            result(FlutterError(code: "INVALID_CONFIGURATION", message: "apiKey is required", details: nil))
            return
        }
        let options = args["options"] as? [String: Any] ?? [:]

        var configuration = SignapConfiguration(apiKey: apiKey)
        if let endpoint = config["endpoint"] as? String, let url = URL(string: endpoint) {
            configuration.endpoint = url
        }
        if let region = config["region"] as? String {
            configuration.region = region
        }
        if let timeoutMs = config["timeoutMs"] as? Double {
            configuration.timeout = timeoutMs / 1000.0
        }
        if let pins = config["pinnedSpkiHashes"] as? [String] {
            configuration.pinnedSPKIHashes = pins
        }

        let identifyOptions = IdentifyOptions(
            linkedId: options["linkedId"] as? String,
            tag: options["tag"] as? String,
            extra: options["extra"] as? [String: String]
        )

        Task {
            do {
                let wise = try Signap(configuration: configuration)
                let r = try await wise.identify(identifyOptions)
                let payload: [String: Any] = [
                    "requestId": r.requestId,
                    "ingestedAt": r.ingestedAt,
                    "region": r.region,
                    "visitorId": r.visitorId,
                    "confidence": r.confidence,
                    "identifiedAt": r.identifiedAt,
                ]
                // FlutterResult must be invoked on the main thread.
                await MainActor.run { result(payload) }
            } catch let error as SignapError {
                let (code, message) = Self.describe(error)
                await MainActor.run { result(FlutterError(code: code, message: message, details: nil)) }
            } catch {
                await MainActor.run {
                    result(FlutterError(code: "NETWORK_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    /// Map the native error taxonomy to the bridge's stable string codes
    /// (mirrors `SignapErrorCode` on the Dart side). Messages carry NO PII.
    private static func describe(_ error: SignapError) -> (String, String) {
        switch error {
        case let .invalidConfiguration(reason):
            return ("INVALID_CONFIGURATION", reason)
        case .network:
            return ("NETWORK_ERROR", "network error")
        case .timeout:
            return ("TIMEOUT", "request timed out")
        case let .http(status):
            return ("HTTP_ERROR", "server returned \(status)")
        case .invalidResponse:
            return ("INVALID_RESPONSE", "invalid response body")
        case .pinningFailed:
            return ("PINNING_FAILED", "certificate pin mismatch")
        }
    }
}

File: android/src/main/kotlin/io/signap/sdk/flutter/SignapPlugin.kt
package io.signap.sdk.flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.signap.sdk.IdentifyOptions
import io.signap.sdk.SignapConfiguration
import io.signap.sdk.SignapException
import io.signap.sdk.Signap
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.Executors

/**
 * Flutter plugin for the Signap Android SDK.
 *
 * Marshals the Dart `identify(config, options)` call to the native
 * `io.signap.sdk.Signap` SDK (sdks/android) — signal collection,
 * derived ids, cert pinning and the `/v1/identify` transport all run there. This
 * plugin only translates the channel `Map` ↔ the native types and runs the
 * blocking `identify()` off the platform thread.
 */
class SignapPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var context: Context? = null

    // identify() does blocking network I/O — keep it off the platform thread.
    private val executor = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
        executor.shutdown()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method != "identify") {
            result.notImplemented()
            return
        }

        @Suppress("UNCHECKED_CAST")
        val configMap = call.argument<Map<String, Any?>>("config")
        @Suppress("UNCHECKED_CAST")
        val optionsMap = call.argument<Map<String, Any?>>("options") ?: emptyMap()
        val ctx = context

        val apiKey = configMap?.get("apiKey") as? String
        if (apiKey == null || ctx == null) {
            result.error("INVALID_CONFIGURATION", "apiKey and a valid context are required", null)
            return
        }

        val configuration = SignapConfiguration(
            apiKey = apiKey,
            endpoint = configMap["endpoint"] as? String,
            region = configMap["region"] as? String ?: "ap",
            timeoutMs = (configMap["timeoutMs"] as? Number)?.toInt() ?: 5000,
            pinnedSpkiHashes = (configMap["pinnedSpkiHashes"] as? List<*>)
                ?.filterIsInstance<String>() ?: emptyList(),
        )

        @Suppress("UNCHECKED_CAST")
        val extra = (optionsMap["extra"] as? Map<*, *>)
            ?.entries
            ?.mapNotNull { (k, v) -> if (k is String && v is String) k to v else null }
            ?.toMap()
        val identifyOptions = IdentifyOptions(
            linkedId = optionsMap["linkedId"] as? String,
            tag = optionsMap["tag"] as? String,
            extra = extra?.ifEmpty { null },
        )

        executor.execute {
            try {
                val signap = Signap(ctx, configuration)
                val r = signap.identify(identifyOptions)
                val payload = mapOf(
                    "requestId" to r.requestId,
                    "ingestedAt" to r.ingestedAt,
                    "region" to r.region,
                    "visitorId" to r.visitorId,
                    "confidence" to r.confidence,
                    "identifiedAt" to r.identifiedAt,
                )
                // Channel replies must be delivered on the main thread.
                mainHandler.post { result.success(payload) }
            } catch (e: SignapException) {
                mainHandler.post { result.error(codeFor(e), e.message, null) }
            } catch (e: Exception) {
                mainHandler.post { result.error("NETWORK_ERROR", e.message, null) }
            }
        }
    }

    private companion object {
        const val CHANNEL = "signap_signals"

        /** Map the native error taxonomy to the bridge's stable string codes
         *  (mirrors `SignapErrorCode` on the Dart side). Messages carry NO PII. */
        fun codeFor(e: SignapException): String = when (e) {
            is SignapException.InvalidConfiguration -> "INVALID_CONFIGURATION"
            SignapException.Network -> "NETWORK_ERROR"
            SignapException.Timeout -> "TIMEOUT"
            is SignapException.Http -> "HTTP_ERROR"
            SignapException.InvalidResponse -> "INVALID_RESPONSE"
            SignapException.PinningFailed -> "PINNING_FAILED"
        }
    }
}
